### PR TITLE
PDPDEVTOOL-5017 - Unit testing stubs 3

### PR DESCRIPTION
**Pull request by vic-nik**

### CODE REVIEW

File: packages/unit-testing/jest-configuration/SuiteCloudJestConfiguration.js [hunk 1]
```diff
@@ -205,10 +205,22 @@ const CORE_STUBS = [
...
+	{
+		module: 'N/keyControl/key',
+		path: `<rootDir>/node_modules/${CORE_STUBS_PATH}/keyControl/Key.js`,
+	},
 	{
 		module: 'N/log',
 		path: `<rootDir>/node_modules/${CORE_STUBS_PATH}/log/log.js`,
 	},
+	{
+		module: 'N/portlet',
```
Comment (lfpvillegas) [inline packages/unit-testing/jest-configuration/SuiteCloudJestConfiguration.js:221]: if we are putting alphabeticly, this should go after plugin

Comment (davidecorreu) [reply packages/unit-testing/jest-configuration/SuiteCloudJestConfiguration.js]: Reordered.

File: packages/unit-testing/stubs/util/util.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+    /**
+     * @memberof util
+     * @name util.deepExtend
+     *
+     * @param {Object} receiver
+     * @param {Object} contributor
+     * @returns {Object} receiver
+     */
+    util.prototype.deepExtend = function (receiver, contributor) { };
```
Comment (karl-anthony-ng) [inline packages/unit-testing/stubs/util/util.js:38]: Not existing in Help.

File: packages/unit-testing/stubs/util/util.js [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    /**
+     * Determines if a variable refers to a number or a string
+     *
+     * @memberof util
+     * @name util.isNumberOrString
+     *
+     * @param obj
+     * @returns {boolean}
+     */
+    util.prototype.isNumberOrString = function (obj) { };
```
Comment (karl-anthony-ng) [inline packages/unit-testing/stubs/util/util.js:126]: Not in Help.

File: packages/unit-testing/stubs/util/util.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    /**
+     * Determines if a variable refers to an Error
+     *
+     * @memberof util
+     * @name util.isError
+     *
+     * @param obj
+     * @returns {boolean}
+     */
+    util.prototype.isError = function (obj) { };
```
Comment (karl-anthony-ng) [inline packages/unit-testing/stubs/util/util.js:160]: Not in Help.

File: packages/unit-testing/stubs/util/util.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+    /**
+     * Remove leading and trailing whitespace from a string
+     *
+     * @memberof util
+     * @name util.trim
+     *
+     * @param {string} str String to have leading/trailing whitespace extracted
+     */
+    util.prototype.trim = function (str) { };
```
Comment (karl-anthony-ng) [inline packages/unit-testing/stubs/util/util.js:170]: Not in Help.

**Review by karl-anthony-ng — CHANGES_REQUESTED** [review]

File: packages/unit-testing/stubs/keyControl/keyControl.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+     * Creates a key object
+     * @restriction Server SuiteScript only
+     * @governance none
+     * @param {Object} options
+     * @param {File} options.file File object containing the key
+     * @param {string} [options.password] Password for the key file (can be in a form of guid token or secret - "custsecret...")
+     * @param {string} [options.scriptId] Desired scriptId under which this key should be later saved
+     * @param {string} options.name Name of the key
+     * @param {string} [options.description] Description for this key
+     * @param {Array<string>} [options.scriptRestrictions] Array of scripts allowed to use the key
```
[the remark this answers is not in the document]
Comment (davidecorreu) [reply packages/unit-testing/stubs/keyControl/keyControl.js:34]: Leaving it as it was.

File: packages/unit-testing/stubs/keyControl/keyControl.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+     */
+    keyControl.prototype.deleteKey = function (options) { }
+
+    /**
+     * Loads a key
+     * @restriction Server SuiteScript only
+     * @governance 10 units
+     * @param {Object} options
+     * @param {string} options.scriptId Id of the key to load
+     * @return {Key}
```
[the remark this answers is not in the document]
Comment (davidecorreu) [reply packages/unit-testing/stubs/keyControl/keyControl.js:60]: Leaving it as it was.

File: packages/unit-testing/stubs/keyControl/Key.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+         * @type {Array<number>}
+         * @since 2019.2
+         */
+        this.restrictions = undefined;
+        /**
+         * @name Key#scriptRestrictions
+         * @type {Array<string>}
+         * @since 2019.2
+         */
+        this.scriptRestrictions = undefined;
```
[the remark this answers is not in the document]
Comment (davidecorreu) [reply packages/unit-testing/stubs/keyControl/Key.js:27]: Leaving it as it was.

File: packages/unit-testing/stubs/sftp/sftp.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+     * @throws {SuiteScriptError} AUTHENTICATION_FAIL_TOO_MANY_INCORRECT_AUTHENTICATION_ATTEMPTS - thrown when number of incorrect authentication attempts is reached
+     * @throws {SuiteScriptError} NO_ROUTE_TO_HOST_FOUND - thrown when host cannot be found on the network
+     * @throws {SuiteScriptError} CONNECTION_RESET - thrown when the connection is closed prematurely
+     * @throws {SuiteScriptError} CONNECTION_CLOSED_BY_HOST - thrown when host does not accept the connection
+     * @throws {SuiteScriptError} THE_REMOTE_PATH_FOR_FILE_IS_NOT_VALID - thrown when base directory does not exist on remote server
+     * @throws {SuiteScriptError} SFTPCREDENTIAL_ENCODING_ERROR - thrown when credentials are incorrectly encoded
+     * @throws {SuiteScriptError} UNABLE_TO_GET_SFTP_SERVER_ADDRESS - thrown when some error occurred while determining remote server address
+     * @throws {SuiteScriptError} SSS_MISSING_REQD_ARGUMENT when any require parameter is missing
+     * @throws {SuiteScriptError} SSS_INVALID_TYPE_ARG when any parameter has incorrect type
```
[the remark this answers is not in the document]
Comment (davidecorreu) [reply packages/unit-testing/stubs/sftp/sftp.js:50]: Leaving it as it was.

File: packages/unit-testing/stubs/task/PivotTaskStatus.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
+define([], function () {
```
[the remark this answers is not in the document]
Comment (davidecorreu) [reply packages/unit-testing/stubs/task/PivotTaskStatus.js:1]: Leaving it as it was.

File: packages/unit-testing/stubs/util/util.js (hunk 2 above)
Comment (davidecorreu) [reply packages/unit-testing/stubs/util/util.js:38]: Leaving it as it was.

File: packages/unit-testing/stubs/util/util.js (hunk 3 above)
Comment (davidecorreu) [reply packages/unit-testing/stubs/util/util.js:126]: Leaving it as it was.

File: packages/unit-testing/stubs/util/util.js (hunk 4 above)
Comment (davidecorreu) [reply packages/unit-testing/stubs/util/util.js:160]: Leaving it as it was.

**Review by lfpvillegas — APPROVED** [review]

**Review by karl-anthony-ng — APPROVED** [review]

**Pull request merged**